<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Fast Unified Random Forests for Survival, Regression, and Classification (RF-SRC) — rfsrc • Fast Unified Random Forests with randomForestSRC</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fast Unified Random Forests for Survival, Regression, and Classification (RF-SRC) — rfsrc"><meta property="og:description" content="

Fast OpenMP-parallel implementation of random forests (Breiman, 2001)
for regression, classification, survival analysis (Ishwaran et al.,
2008), competing risks (Ishwaran et al., 2012), multivariate outcomes
(Segal and Xiao, 2011), unsupervised learning (Mantero and Ishwaran,
2020), quantile regression (Meinshausen, 2006; Zhang et al., 2019;
Greenwald and Khanna, 2001), and imbalanced q-classification (O'Brien
and Ishwaran, 2019).

The package supports both deterministic and randomized splitting rules
(Geurts et al., 2006; Ishwaran, 2015) across all families. Multiple
types of variable importance (VIMP) are available, including holdout
VIMP and confidence regions (Ishwaran and Lu, 2019), for both individual
and grouped variables. Variable selection can be performed using minimal
depth (Ishwaran et al., 2010, 2011). Fast interfaces for missing data
imputation are provided using several forest-based algorithms (Tang and
Ishwaran, 2017).

Highlighted updates:

For survival and competing risk analysis, concordance-based
  performance is now computed using Uno inverse-probability-of-censoring
  weighting (Uno et al. 2001). This affects all survival performance
  values derived from the concordance index, including out-of-bag and
  test error rates and variable importance (VIMP). To revert to
  unweighted Harrell concordance, set use.uno = FALSE when
  fitting a survival forest.  Concordance calculations now use an
  efficient \(O(n \log n)\) algorithm based on a binary indexed tree
  (Fenwick 1994, Therneau 2024), replacing the naive \(O(n^2)\)
  pairwise computation for large \(n\).
For variable selection, we recommend using VarPro, an R
  package for model-independent variable selection using rule-based
  variable priority. It supports regression, classification, survival
  analysis, and includes a new mode for unsupervised learning. See
  https://www.varprotools.org for more information.
For computational speed, the default VIMP method has changed
  from &quot;permute&quot; (Breiman-Cutler permutation) to &quot;anti&quot;
  (importance = &quot;anti&quot; or importance = TRUE). While
  faster, this may be less accurate in settings such as highly
  imbalanced classification. To revert to permutation VIMP, use
  importance = &quot;permute&quot;.



This is the main entry point to the randomForestSRC
package. For more information on OpenMP support and the package as a
whole, see package?randomForestSRC."><meta property="og:image" content="https://www.randomforestsrc.org/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-WZWXVK817Z"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-WZWXVK817Z');
</script></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Fast Unified Random Forests with randomForestSRC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/getstarted.html">
    <span class="fas fa-home fa-lg"></span>

    Getting Started
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../cheatsheets.pdf">Cheatsheet</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/parallel.html">Parallel Processing</a>
    </li>
    <li>
      <a href="../articles/hybrid.html">Hybrid Parallel Processing</a>
    </li>
    <li>
      <a href="../articles/algorithm.html">randomForestSRC Algorithm</a>
    </li>
    <li>
      <a href="../articles/speedup.html">Speedup Random Forest Analyses</a>
    </li>
    <li>
      <a href="../articles/forestWgt.html">Forest Weights, In-Bag (IB) and Out-of-Bag (OOB) Ensembles</a>
    </li>
    <li>
      <a href="../articles/survival.html">Random Survival Forests</a>
    </li>
    <li>
      <a href="../articles/competing.html">Competing Risks</a>
    </li>
    <li>
      <a href="../articles/imbalance.html">Random Forests Quantile Classifier (RFQ)</a>
    </li>
    <li>
      <a href="../articles/mvsplit.html">Multivariate Splitting Rule</a>
    </li>
    <li>
      <a href="../articles/aucsplit.html">AUC Splitting Rule for Multiclass Problems</a>
    </li>
    <li>
      <a href="../articles/sidClustering.html">sidClustering</a>
    </li>
    <li>
      <a href="../articles/rfsrc-subsample.html">Variable Importance (VIMP) with Subsampling Inference</a>
    </li>
    <li>
      <a href="../articles/minidep.html">Minimal Depth</a>
    </li>
    <li>
      <a href="../articles/partial.html">Partial Plots</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li>
  <a href="https://github.com/kogalur/randomForestSRC/issues" class="external-link">Bug Reporting</a>
</li>
<li>
  <a href="https://github.com/kogalur/randomForestSRC/discussions" class="external-link">Discussions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://kogalur.github.io/randomForestSRC/javadoc" class="external-link">Java API</a>
</li>
<li>
  <a href="https://github.com/kogalur/randomForestSRC/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

    GitHub
  </a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kogalur/randomForestSRC/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

    GitHub
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fast Unified Random Forests for Survival, Regression, and Classification (RF-SRC)</h1>

    <div class="hidden name"><code>rfsrc.Rd</code></div>
    </div>

    <div class="ref-description">


<p>Fast OpenMP-parallel implementation of random forests (Breiman, 2001)
for regression, classification, survival analysis (Ishwaran et al.,
2008), competing risks (Ishwaran et al., 2012), multivariate outcomes
(Segal and Xiao, 2011), unsupervised learning (Mantero and Ishwaran,
2020), quantile regression (Meinshausen, 2006; Zhang et al., 2019;
Greenwald and Khanna, 2001), and imbalanced q-classification (O'Brien
and Ishwaran, 2019).</p>

<p>The package supports both deterministic and randomized splitting rules
(Geurts et al., 2006; Ishwaran, 2015) across all families. Multiple
types of variable importance (VIMP) are available, including holdout
VIMP and confidence regions (Ishwaran and Lu, 2019), for both individual
and grouped variables. Variable selection can be performed using minimal
depth (Ishwaran et al., 2010, 2011). Fast interfaces for missing data
imputation are provided using several forest-based algorithms (Tang and
Ishwaran, 2017).</p>

<p><strong>Highlighted updates:</strong></p>
<ol><li><p>For survival and competing risk analysis, concordance-based
  performance is now computed using Uno inverse-probability-of-censoring
  weighting (Uno et al. 2001). This affects all survival performance
  values derived from the concordance index, including out-of-bag and
  test error rates and variable importance (VIMP). To revert to
  unweighted Harrell concordance, set <code>use.uno = FALSE</code> when
  fitting a survival forest.  Concordance calculations now use an
  efficient \(O(n \log n)\) algorithm based on a binary indexed tree
  (Fenwick 1994, Therneau 2024), replacing the naive \(O(n^2)\)
  pairwise computation for large \(n\).</p></li>
<li><p>For variable selection, we recommend using <span class="pkg">VarPro</span>, an R
  package for model-independent variable selection using rule-based
  variable priority. It supports regression, classification, survival
  analysis, and includes a new mode for unsupervised learning. See
  <a href="https://www.varprotools.org" class="external-link">https://www.varprotools.org</a> for more information.</p></li>
<li><p>For computational speed, the default VIMP method has changed
  from "permute" (Breiman-Cutler permutation) to "anti"
  (<code>importance = "anti"</code> or <code>importance = TRUE</code>). While
  faster, this may be less accurate in settings such as highly
  imbalanced classification. To revert to permutation VIMP, use
  <code>importance = "permute"</code>.</p></li>
</ol><p>This is the main entry point to the <span class="pkg">randomForestSRC</span>
package. For more information on OpenMP support and the package as a
whole, see <code><a href="randomForestSRC_package.html">package?randomForestSRC</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rfsrc</span><span class="op">(</span><span class="va">formula</span>, <span class="va">data</span>, ntree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  mtry <span class="op">=</span> <span class="cn">NULL</span>, ytry <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nodesize <span class="op">=</span> <span class="cn">NULL</span>, nodedepth <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitrule <span class="op">=</span> <span class="cn">NULL</span>, nsplit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  importance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="st">"none"</span>, <span class="st">"anti"</span>, <span class="st">"permute"</span>, <span class="st">"random"</span><span class="op">)</span>,</span>
<span>  block.size <span class="op">=</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">is.element</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">importance</span><span class="op">)</span>,</span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"FALSE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="cn">NULL</span> <span class="kw">else</span> <span class="fl">10</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"by.root"</span>, <span class="st">"none"</span>, <span class="st">"by.user"</span><span class="op">)</span>,</span>
<span>  samptype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"swor"</span>, <span class="st">"swr"</span><span class="op">)</span>, samp <span class="op">=</span> <span class="cn">NULL</span>, membership <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sampsize <span class="op">=</span> <span class="kw">if</span> <span class="op">(</span><span class="va">samptype</span> <span class="op">==</span> <span class="st">"swor"</span><span class="op">)</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span> <span class="op">*</span> <span class="fl">.632</span><span class="op">}</span> <span class="kw">else</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">}</span>,</span>
<span>  na.action <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"na.omit"</span>, <span class="st">"na.impute"</span><span class="op">)</span>, nimpute <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ntime <span class="op">=</span> <span class="fl">150</span>, <span class="va">cause</span>,</span>
<span>  perf.type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  proximity <span class="op">=</span> <span class="cn">FALSE</span>, distance <span class="op">=</span> <span class="cn">FALSE</span>, forest.wt <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  xvar.wt <span class="op">=</span> <span class="cn">NULL</span>, yvar.wt <span class="op">=</span> <span class="cn">NULL</span>, split.wt <span class="op">=</span> <span class="cn">NULL</span>, case.wt  <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  case.depth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  forest <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use.uno <span class="op">=</span> <span class="cn">TRUE</span>, save.memory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  var.used <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="st">"all.trees"</span>, <span class="st">"by.tree"</span><span class="op">)</span>,</span>
<span>  split.depth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="st">"all.trees"</span>, <span class="st">"by.tree"</span><span class="op">)</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  do.trace <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## convenient interface for growing a CART tree</span></span>
<span><span class="fu">rfsrc.cart</span><span class="op">(</span><span class="va">formula</span>, <span class="va">data</span>, ntree <span class="op">=</span> <span class="fl">1</span>, mtry <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>,</span>
<span>   bootstrap <span class="op">=</span> <span class="st">"none"</span>, nsplit <span class="op">=</span> <span class="fl">0</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<p></p>
<dl><dt id="arg-formula">formula<a class="anchor" aria-label="anchor" href="#arg-formula"></a></dt>
<dd><p>A formula describing the model to fit. Interaction terms are not supported. If missing, unsupervised splitting is used.</p></dd>


  <dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>Data frame containing the response and predictor variables.</p></dd>


  <dt id="arg-ntree">ntree<a class="anchor" aria-label="anchor" href="#arg-ntree"></a></dt>
<dd><p>Number of trees to grow.</p></dd>


  <dt id="arg-mtry">mtry<a class="anchor" aria-label="anchor" href="#arg-mtry"></a></dt>
<dd><p>Number of candidate variables randomly selected at each split. Defaults: regression uses <code>p/3</code>, others use <code>sqrt(p)</code>; rounded up.</p></dd>


  <dt id="arg-ytry">ytry<a class="anchor" aria-label="anchor" href="#arg-ytry"></a></dt>
<dd><p>Number of pseudo-response variables randomly selected for unsupervised splitting. Default is 1.</p></dd>


  <dt id="arg-nodesize">nodesize<a class="anchor" aria-label="anchor" href="#arg-nodesize"></a></dt>
<dd><p>Minimum terminal node size. Defaults: survival/competing risks (15), regression (5), classification (1), mixed/unsupervised (3).</p></dd>


  <dt id="arg-nodedepth">nodedepth<a class="anchor" aria-label="anchor" href="#arg-nodedepth"></a></dt>
<dd><p>Maximum tree depth. Ignored by default.</p></dd>


  <dt id="arg-splitrule">splitrule<a class="anchor" aria-label="anchor" href="#arg-splitrule"></a></dt>
<dd><p>Splitting rule. See Details.</p></dd>


  <dt id="arg-nsplit">nsplit<a class="anchor" aria-label="anchor" href="#arg-nsplit"></a></dt>
<dd><p>Number of random split points per variable. <code>0</code> uses all values (deterministic). Default is 10.</p></dd>


  <dt id="arg-importance">importance<a class="anchor" aria-label="anchor" href="#arg-importance"></a></dt>
<dd><p>Variable importance (VIMP) method. Choices: <code>FALSE</code>, <code>TRUE</code>, <code>"none"</code>, <code>"anti"</code>, <code>"permute"</code>, <code>"random"</code>. Default is <code>"none"</code>. VIMP can be computed later using <code>vimp</code> or <code>predict</code>.</p></dd>


  <dt id="arg-block-size">block.size<a class="anchor" aria-label="anchor" href="#arg-block-size"></a></dt>
<dd><p>Controls frequency of cumulative error/VIMP updates. Default is <code>10</code> if importance is requested; otherwise <code>NULL</code>. See Details.</p></dd>


  <dt id="arg-bootstrap">bootstrap<a class="anchor" aria-label="anchor" href="#arg-bootstrap"></a></dt>
<dd><p>Bootstrap method. Options: <code>"by.root"</code> (default), <code>"by.user"</code>, or <code>"none"</code> (no OOB error possible).</p></dd>


  <dt id="arg-samptype">samptype<a class="anchor" aria-label="anchor" href="#arg-samptype"></a></dt>
<dd><p>Sampling type for <code>by.root</code> bootstrap. Options: <code>"swor"</code> (without replacement, default), <code>"swr"</code> (with replacement).</p></dd>


  <dt id="arg-samp">samp<a class="anchor" aria-label="anchor" href="#arg-samp"></a></dt>
<dd><p>Bootstrap weights (only for <code>bootstrap="by.user"</code>). A matrix of size <code>n x ntree</code> giving in-bag counts per tree.</p></dd>


  <dt id="arg-membership">membership<a class="anchor" aria-label="anchor" href="#arg-membership"></a></dt>
<dd><p>Return inbag and terminal node membership?</p></dd>


  <dt id="arg-sampsize">sampsize<a class="anchor" aria-label="anchor" href="#arg-sampsize"></a></dt>
<dd><p>Bootstrap sample size (used when <code>bootstrap="by.root"</code>). Defaults: 0.632 x n for <code>swor</code>; n for <code>swr</code>. Can also be numeric.</p></dd>


  <dt id="arg-na-action">na.action<a class="anchor" aria-label="anchor" href="#arg-na-action"></a></dt>
<dd><p>Missing data handling. <code>"na.omit"</code> (default) removes rows with any <code>NA</code>; <code>"na.impute"</code> performs fast internal imputation. See also <code>impute</code>.</p></dd>


  <dt id="arg-nimpute">nimpute<a class="anchor" aria-label="anchor" href="#arg-nimpute"></a></dt>
<dd><p>Number of iterations for internal imputation. If &gt;1, OOB error rates may be optimistic.</p></dd>


  <dt id="arg-ntime">ntime<a class="anchor" aria-label="anchor" href="#arg-ntime"></a></dt>
<dd><p>For survival models: number or grid of time points used in ensemble estimation. If <code>NULL</code> or <code>0</code>, uses all event times.</p></dd>


  <dt id="arg-cause">cause<a class="anchor" aria-label="anchor" href="#arg-cause"></a></dt>
<dd><p>For competing risks: event of interest (1 to <code>J</code>), or a vector of weights over all <code>J</code> events. Defaults to an average over all events.</p></dd>


  <dt id="arg-perf-type">perf.type<a class="anchor" aria-label="anchor" href="#arg-perf-type"></a></dt>
<dd><p>Optional performance metric for prediction, VIMP, and error. Defaults to the family-specific metric. <code>"none"</code> disables performance. See Details.</p></dd>


  <dt id="arg-proximity">proximity<a class="anchor" aria-label="anchor" href="#arg-proximity"></a></dt>
<dd><p>Compute proximity matrix? Options: <code>"inbag"</code>, <code>"oob"</code>, <code>"all"</code>, <code>TRUE</code> (inbag), or <code>FALSE</code>.</p></dd>


  <dt id="arg-distance">distance<a class="anchor" aria-label="anchor" href="#arg-distance"></a></dt>
<dd><p>Compute pairwise distances between cases? Similar options as <code>proximity</code>. See Details.</p></dd>


  <dt id="arg-forest-wt">forest.wt<a class="anchor" aria-label="anchor" href="#arg-forest-wt"></a></dt>
<dd><p>Return forest weight matrix? Same options as <code>proximity</code>. Default is <code>TRUE</code> (inbag).</p></dd>


  <dt id="arg-xvar-wt">xvar.wt<a class="anchor" aria-label="anchor" href="#arg-xvar-wt"></a></dt>
<dd><p>Optional weights on x-variables for sampling at splits. Does not need to sum to 1. Defaults to uniform.</p></dd>


  <dt id="arg-yvar-wt">yvar.wt<a class="anchor" aria-label="anchor" href="#arg-yvar-wt"></a></dt>
<dd><p>Weights on response variables (for multivariate regression). Used when <code>y</code> is high-dimensional.</p></dd>


  <dt id="arg-split-wt">split.wt<a class="anchor" aria-label="anchor" href="#arg-split-wt"></a></dt>
<dd><p>Weights applied to each variable's split statistic. Higher weight increases likelihood of splitting.</p></dd>


  <dt id="arg-case-wt">case.wt<a class="anchor" aria-label="anchor" href="#arg-case-wt"></a></dt>
<dd><p>Sampling weights for cases in the bootstrap. Higher values increase selection probability. See class imbalance example.</p></dd>


  <dt id="arg-case-depth">case.depth<a class="anchor" aria-label="anchor" href="#arg-case-depth"></a></dt>
<dd><p>Return matrix recording depth of first split for each case? Default is <code>FALSE</code>.</p></dd>


  <dt id="arg-forest">forest<a class="anchor" aria-label="anchor" href="#arg-forest"></a></dt>
<dd><p>Save forest object for future prediction? Set <code>FALSE</code> if prediction is not needed.</p></dd>


  <dt id="arg-use-uno">use.uno<a class="anchor" aria-label="anchor" href="#arg-use-uno"></a></dt>
<dd><p>Logical.  If <code>TRUE</code> (default for survival), concordance-based performance is computed using inverse-probability-of-censoring weights as
  proposed by Uno et al. (2011). This affects all survival performance values derived from the C-index, including error rates (C-error) and variable
  importance (VIMP). Set <code>use.uno = FALSE</code> to use the unweighted Harrell concordance. Applies only to survival families.</p></dd>


  <dt id="arg-save-memory">save.memory<a class="anchor" aria-label="anchor" href="#arg-save-memory"></a></dt>
<dd><p>Reduce memory usage by avoiding storage of prediction quantities. Recommended for large survival or competing risk forests.</p></dd>


  <dt id="arg-var-used">var.used<a class="anchor" aria-label="anchor" href="#arg-var-used"></a></dt>
<dd><p>Return variable usage statistics? Options: <code>FALSE</code>, <code>"all.trees"</code>, <code>"by.tree"</code>.</p></dd>


  <dt id="arg-split-depth">split.depth<a class="anchor" aria-label="anchor" href="#arg-split-depth"></a></dt>
<dd><p>Return minimal depth of splits for each variable? Options: <code>FALSE</code>, <code>"all.trees"</code>, <code>"by.tree"</code>. See Details.</p></dd>


  <dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Integer seed for reproducibility (negative values only).</p></dd>


  <dt id="arg-do-trace">do.trace<a class="anchor" aria-label="anchor" href="#arg-do-trace"></a></dt>
<dd><p>Print progress updates every <code>do.trace</code> seconds.</p></dd>


  <dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to or from other methods.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p></p><ol><li><p><em>Types of forests</em></p>
<p>The type of forest is automatically inferred from the outcome and formula. Supported forest types include:</p>
<p></p><ul><li><p>Regression forests for continuous outcomes.</p></li>
<li><p>Classification forests for factor outcomes.</p></li>
<li><p>Multivariate forests for continuous, categorical, or mixed outcomes.</p></li>
<li><p>Unsupervised forests when no outcome is specified.</p></li>
<li><p>Survival forests for right-censored time-to-event data.</p></li>
<li><p>Competing risk forests for multi-event survival settings.</p></li>
</ul></li>
<li><p><em>Splitting</em></p>
<p></p><ol><li><p>Splitting rules are set using the <code>splitrule</code> option.</p></li>
<li><p>Random splitting is invoked via <code>splitrule = "random"</code>.</p></li>
<li><p>Use <code>nsplit</code> to enable randomized splitting and improve speed; see <em>Improving computational speed</em>.</p></li>
</ol></li>
<li><p><em>Available splitting rules</em></p>
<p></p><ul><li><p><strong>Regression</strong></p><ol><li><p><code>"mse"</code> (default): weighted mean squared error (Breiman et al., 1984).</p></li>
<li><p><code>"quantile.regr"</code>: quantile regression via check-loss; see <code>quantreg.rfsrc</code>.</p></li>
<li><p><code>"la.quantile.regr"</code>: local adaptive quantile regression.</p></li>
</ol></li>
<li><p><strong>Classification</strong></p><ol><li><p><code>"gini"</code> (default): Gini index.</p></li>
<li><p><code>"auc"</code>: AUC-based splitting; appropriate for imbalanced data.</p></li>
<li><p><code>"entropy"</code>: entropy-based splitting.</p></li>
</ol></li>
<li><p><strong>Survival</strong></p><ol><li><p><code>"logrank"</code> (default): log-rank splitting.</p></li>
<li><p><code>"bs.gradient"</code>: Brier score gradient splitting. Uses 90th percentile of observed times by default, or set <code>prob</code>.</p></li>
<li><p><code>"logrankscore"</code>: log-rank score splitting.</p></li>
</ol></li>
<li><p><strong>Competing risks</strong> (see Ishwaran et al., 2014)</p><ol><li><p><code>"logrankCR"</code> (default): Gray's test-based weighted log-rank splitting.</p></li>
<li><p><code>"logrank"</code>: cause-specific weighted log-rank; use <code>cause</code> to target specific events.</p></li>
</ol></li>
<li><p><strong>Multivariate</strong></p><ol><li><p>Default: normalized composite splitting (Tang and Ishwaran, 2017).</p></li>
<li><p><code>"mahalanobis"</code>: Mahalanobis splitting with optional covariance matrix; for multivariate regression.</p></li>
</ol></li>
<li><p><strong>Unsupervised</strong>
    Splitting uses pseudo-outcomes and the composite rule. See <code>sidClustering</code> for advanced unsupervised analysis.</p></li>
<li><p><strong>Custom splitting</strong>
    Custom rules can be defined using <code>splitCustom.c</code>. Up to 16 rules per family are allowed. Use <code>"custom"</code>, <code>"custom1"</code>, etc. Compilation required.</p></li>
</ul></li>
<li><p><em>Improving computational speed</em></p>
<p>See <code><a href="rfsrc.fast.html">rfsrc.fast</a></code>. Strategies include:</p>
<p></p><ul><li><p>Increase <code>nodesize</code>.</p></li>
<li><p>Set <code>save.memory = TRUE</code> for large survival or competing risk models.</p></li>
<li><p>Set <code>block.size = NULL</code> to avoid repeated cumulative error computation.</p></li>
<li><p>Use <code>perf.type = "none"</code> to disable VIMP and C-index calculations.</p></li>
<li><p>Set <code>nsplit</code> to a small integer (e.g., 1-10).</p></li>
<li><p>Reduce bootstrap size with <code>sampsize</code>, <code>samptype</code>.</p></li>
<li><p>Set <code>ntime</code> to a coarse grid (e.g., 50) for survival models.</p></li>
<li><p>Pre-filter variables; use <code>max.subtree</code> for fast variable selection.</p></li>
</ul></li>
<li><p><em>Prediction Error</em></p>
<p>Error is computed using OOB data:</p><ul><li><p>Regression: mean squared error.</p></li>
<li><p>Classification: misclassification rate, Brier score, AUC.</p></li>
<li><p>Survival: C-error = 1 - Harrell's concordance index.</p></li>
</ul><p>If <code>bootstrap = "none"</code>, OOB error is unavailable. Use <code>predict.rfsrc</code> for cross-validation error instead.</p></li>
<li><p><em>Variable Importance (VIMP)</em></p>
<p>VIMP methods:</p><ul><li><p><code>"permute"</code>: permutation VIMP (Breiman-Cutler).</p></li>
<li><p><code>"random"</code>: randomized left/right assignment.</p></li>
<li><p><code>"anti"</code> (default): anti-split assignment.</p></li>
</ul><p>The <code>block.size</code> option controls granularity. For confidence intervals, see <code>subsampling</code>. Also see <code>holdout.vimp</code> for a more conservative variant.</p></li>
<li><p><em>Multivariate Forests</em></p>
<p>Use:</p><div class="sourceCode"><pre><code><span><span class="fu"><a href="../reference/rfsrc.html">rfsrc</a></span><span class="op">(</span><span class="fu">Multivar</span><span class="op">(</span><span class="va">y1</span>, <span class="va">...</span>, <span class="va">yd</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">data</span><span class="op">)</span></span></code></pre></div><p>or</p><div class="sourceCode"><pre><code><span><span class="fu"><a href="../reference/rfsrc.html">rfsrc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">y1</span>, <span class="va">...</span>, <span class="va">yd</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">data</span><span class="op">)</span></span></code></pre></div>

<p>Use <code>get.mv.formula</code>, <code>get.mv.predicted</code>, <code>get.mv.error</code> for multivariate extraction.</p></li>
<li><p><em>Unsupervised Forests</em></p>
<p>Use:</p><div class="sourceCode"><pre><code><span><span class="fu"><a href="../reference/rfsrc.html">rfsrc</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">X</span><span class="op">)</span></span></code></pre></div><p>or</p><div class="sourceCode"><pre><code><span><span class="fu"><a href="../reference/rfsrc.html">rfsrc</a></span><span class="op">(</span><span class="fu">Unsupervised</span><span class="op">(</span><span class="va">ytry</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">X</span><span class="op">)</span></span></code></pre></div>

<p>Random subsets of <code>ytry</code> pseudo-responses are used for each <code>mtry</code> variable. No performance metrics are computed.</p></li>
<li><p><em>Survival, Competing Risks</em></p>
<p></p><ul><li><p>Survival: use <code>Surv(time, status) ~ .</code>. Status must be 0 (censored) or 1 (event).</p></li>
<li><p>Competing risks: status = 0 (censored), 1-J (event types). Use <code>cause</code> to target specific events.</p></li>
<li><p>Larger <code>nodesize</code> is typically needed for competing risks.</p></li>
</ul></li>
<li><p><em>Missing data imputation</em></p>
<p>Use <code>na.action = "na.impute"</code>. Iteration with <code>nimpute &gt; 1</code> replaces missing values using OOB predictions. Observations or variables with all missing values are removed.</p></li>
<li><p><em>Allowable data types and factors</em></p>
<p>Variables must be numeric, integer, factor, or logical. Non-factors are coerced to numeric. For unordered factors, all complementary subsets are considered for splits.</p>
<p>Factor levels are mapped to ensure consistency across training/test data. Consider converting factors to numeric for high-dimensional settings.</p></li>
</ol></div>
    <div id="value">
    <h2>Value</h2>
    <p>An object of class <code>(rfsrc, grow)</code> with the following components:</p>
<dl><dt>call</dt>
<dd><p>The original call to <code>rfsrc</code>.</p></dd>

  <dt>family</dt>
<dd><p>The family used in the analysis.</p></dd>

  <dt>n</dt>
<dd><p>Sample size after applying <code>na.action</code>.</p></dd>

  <dt>ntree</dt>
<dd><p>Number of trees grown.</p></dd>

  <dt>mtry</dt>
<dd><p>Number of variables randomly selected at each node.</p></dd>

  <dt>nodesize</dt>
<dd><p>Minimum terminal node size.</p></dd>

  <dt>nodedepth</dt>
<dd><p>Maximum depth allowed for each tree.</p></dd>

  <dt>splitrule</dt>
<dd><p>Splitting rule used.</p></dd>

  <dt>nsplit</dt>
<dd><p>Number of random split points.</p></dd>

  <dt>yvar</dt>
<dd><p>Response values.</p></dd>

  <dt>yvar.names</dt>
<dd><p>Character vector of response variable names.</p></dd>

  <dt>xvar</dt>
<dd><p>Data frame of predictor variables.</p></dd>

  <dt>xvar.names</dt>
<dd><p>Character vector of predictor variable names.</p></dd>

  <dt>xvar.wt</dt>
<dd><p>Non-negative weights specifying the selection probability of each variable.</p></dd>

  <dt>split.wt</dt>
<dd><p>Non-negative weights adjusting each variable's split statistic.</p></dd>

  <dt>cause.wt</dt>
<dd><p>Weights for composite competing risk splitting.</p></dd>

  <dt>leaf.count</dt>
<dd><p>Number of terminal nodes per tree. A value of 0 indicates a rejected tree (may occur with missing data); a value of 1 indicates a stump.</p></dd>

  <dt>proximity</dt>
<dd><p>Proximity matrix indicating how often case pairs fall in the same terminal node.</p></dd>

  <dt>forest</dt>
<dd><p>Forest object, returned if <code>forest=TRUE</code>. Required for prediction and most wrappers.</p></dd>

  <dt>forest.wt</dt>
<dd><p>Forest weight matrix.</p></dd>

  <dt>membership</dt>
<dd><p>Terminal node membership matrix (rows: trees; columns: cases).</p></dd>

  <dt>inbag</dt>
<dd><p>Inbag count matrix (rows: trees; columns: cases).</p></dd>

  <dt>var.used</dt>
<dd><p>Number of times each variable is used to split a node.</p></dd>

  <dt>imputed.indv</dt>
<dd><p>Indices of individuals with missing values.</p></dd>

  <dt>imputed.data</dt>
<dd><p>Imputed dataset with responses followed by predictors.</p></dd>

  <dt>split.depth</dt>
<dd><p>Matrix or array recording minimal split depth of variables by case and tree.</p></dd>

  <dt>err.rate</dt>
<dd><p>Cumulative OOB error rate.</p></dd>

  <dt>err.block.rate</dt>
<dd><p>Cumulative error per ensemble block (size defined by <code>block.size</code>). If <code>block.size = 1</code>, error per tree.</p></dd>

  <dt>importance</dt>
<dd><p>Variable importance (VIMP) for each predictor.</p></dd>

  <dt>predicted</dt>
<dd><p>In-bag predicted values.</p></dd>

  <dt>predicted.oob</dt>
<dd><p>Out-of-bag (OOB) predicted values.</p></dd>


  <dt>class</dt>
<dd><p>(Classification) In-bag predicted class labels.</p></dd>

  <dt>class.oob</dt>
<dd><p>(Classification) OOB predicted class labels.</p></dd>


  <dt>regrOutput</dt>
<dd><p>(Multivariate) List of performance results for continuous outcomes.</p></dd>

  <dt>clasOutput</dt>
<dd><p>(Multivariate) List of performance results for categorical outcomes.</p></dd>


  <dt>survival</dt>
<dd><p>(Survival) In-bag survival functions.</p></dd>

  <dt>survival.oob</dt>
<dd><p>(Survival) OOB survival functions.</p></dd>

  <dt>chf</dt>
<dd><p>(Survival or competing risks) In-bag cumulative hazard function.</p></dd>

  <dt>chf.oob</dt>
<dd><p>(Survival or competing risks) OOB cumulative hazard function.</p></dd>

  <dt>time.interest</dt>
<dd><p>(Survival or competing risks) Unique sorted event times.</p></dd>

  <dt>ndead</dt>
<dd><p>(Survival or competing risks) Total number of observed events.</p></dd>


  <dt>cif</dt>
<dd><p>(Competing risks) In-bag cumulative incidence function by cause.</p></dd>

  <dt>cif.oob</dt>
<dd><p>(Competing risks) OOB cumulative incidence function by cause.</p></dd>


</dl></div>
    <div id="note">
    <h2>Note</h2>
    <p>Values returned by the forest depend on the family:</p>
<ul><li><p><strong>Regression</strong>: <code>predicted</code> and <code>predicted.oob</code> are vectors of predicted values.</p></li>
<li><p><strong>Classification</strong>: <code>predicted</code> and <code>predicted.oob</code> are matrices of class probabilities. VIMP and performance metrics are returned as a matrix with <code>J+1</code> columns (J = number of classes). The first column ("all") gives unconditional results; remaining columns give class-conditional results.</p></li>
<li><p><strong>Survival</strong>: <code>predicted</code> contains mortality estimates (Ishwaran et al., 2008). These are calibrated to the number of expected events under identical covariate profiles. Also returned are matrices of the survival function and CHF for each individual over <code>time.interest</code>.</p></li>
<li><p><strong>Competing risks</strong>: <code>predicted</code> contains expected life years lost by cause (Ishwaran et al., 2013). Also returned are three-dimensional arrays for CIF and CSCHF indexed by case, time, and event type.</p></li>
<li><p><strong>Multivariate</strong>: Predictions, VIMP, and error rates are returned in <code>regrOutput</code> and <code>clasOutput</code>. Use <code>get.mv.predicted</code>, <code>get.mv.vimp</code>, and <code>get.mv.error</code> to extract results.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Hemant Ishwaran and Udaya B. Kogalur</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Breiman L., Friedman J.H., Olshen R.A. and Stone C.J. (1984).
  <em>Classification and Regression Trees</em>, Belmont, California.</p>
<p>Breiman L. (2001). Random forests, <em>Machine Learning</em>, 45:5-32.</p>
<p>Cutler A. and Zhao G. (2001). PERT-Perfect random tree ensembles.
  <em>Comp. Sci. Statist.</em>, 33: 490-497.</p>
<p>Dietterich, T. G. (2000). An experimental comparison of three methods for
  constructing ensembles of decision trees: bagging, boosting, and randomization.
  <em>Machine Learning</em>, 40, 139-157.</p>
<p>Fenwick, P.M. (1994).  A new data structure for cumulative frequency
  tables.  <em>Software: Practice and Experience</em>, 24(3):327–336.</p>
<p>Gray R.J. (1988).  A class of k-sample tests for comparing the
  cumulative incidence of a competing risk, <em>Ann. Statist.</em>,
  16: 1141-1154.</p>
<p>Geurts, P., Ernst, D. and Wehenkel, L., (2006). Extremely randomized
  trees. <em>Machine learning</em>, 63(1):3-42.</p>
<p>Greenwald M. and Khanna S. (2001).  Space-efficient online computation of
  quantile summaries. <em>Proceedings of ACM SIGMOD</em>, 30(2):58-66.</p>
<p>Harrell et al. F.E. (1982).  Evaluating the yield of medical tests,
  <em>J. Amer. Med. Assoc.</em>, 247:2543-2546.</p>
<p>Hothorn T. and Lausen B. (2003). On the exact distribution of maximally selected
  rank statistics, <em>Comp. Statist. Data Anal.</em>, 43:121-137.</p>
<p>Ishwaran H. (2007).  Variable importance in binary regression
  trees and forests,  <em>Electronic J. Statist.</em>, 1:519-537.</p>
<p>Ishwaran H. and Kogalur U.B. (2007).  Random survival forests for R,
  <em>Rnews</em>, 7(2):25-31.</p>
<p>Ishwaran H., Kogalur U.B., Blackstone E.H. and Lauer M.S.
  (2008).  Random survival forests, <em>Ann. App.
    Statist.</em>, 2:841-860.</p>
<p>Ishwaran H., Kogalur U.B., Gorodeski E.Z, Minn A.J. and
  Lauer M.S. (2010).  High-dimensional variable selection for survival
  data.  <em>J. Amer. Statist. Assoc.</em>, 105:205-217.</p>
<p>Ishwaran H., Kogalur U.B., Chen X. and Minn A.J. (2011). Random survival
  forests for high-dimensional data. <em>Stat. Anal. Data Mining</em>, 4:115-132</p>
<p>Ishwaran H., Gerds T.A., Kogalur U.B., Moore R.D., Gange S.J. and Lau
  B.M. (2014). Random survival forests for competing risks.
  <em>Biostatistics</em>, 15(4):757-773.</p>
<p>Ishwaran H. and Malley J.D. (2014). Synthetic learning
  machines. <em>BioData Mining</em>, 7:28.</p>
<p>Ishwaran H. (2015).  The effect of splitting on random forests.
  <em>Machine Learning</em>, 99:75-118.</p>
<p>Lin, Y. and Jeon, Y. (2006). Random forests and adaptive nearest
  neighbors. <em>J. Amer. Statist. Assoc.</em>, 101(474), 578-590.</p>
<p>Lu M., Sadiq S., Feaster D.J. and Ishwaran H. (2018). Estimating
  individual treatment effect in observational data using random forest
  methods. <em>J. Comp. Graph. Statist</em>, 27(1), 209-219</p>
<p>Ishwaran H. and Lu M.  (2019).  Standard errors and confidence
  intervals for variable importance in random forest regression,
  classification, and survival. <em>Statistics in Medicine</em>, 38,
  558-582.</p>
<p>LeBlanc M. and Crowley J. (1993).  Survival trees by goodness of split,
  <em>J. Amer. Statist. Assoc.</em>, 88:457-467.</p>
<p>Loh W.-Y and Shih Y.-S (1997).  Split selection methods for
  classification trees, <em>Statist. Sinica</em>, 7:815-840.</p>
<p>Mantero A. and Ishwaran H. (2021).  Unsupervised random forests.
  <em>Statistical Analysis and Data Mining</em>, 14(2):144-167.</p>
<p>Meinshausen N. (2006) Quantile regression forests, <em>Journal of
  Machine Learning Research</em>, 7:983-999.</p>
<p>Mogensen, U.B, Ishwaran H. and Gerds T.A. (2012). Evaluating random
  forests for survival analysis using prediction error curves,
  <em>J.  Statist. Software</em>, 50(11): 1-23.</p>
<p>O'Brien R. and Ishwaran H. (2019).  A random forests quantile
  classifier for class imbalanced data. <em>Pattern Recognition</em>,
  90, 232-249</p>
<p>Segal M.R. (1988).  Regression trees for censored data,
  <em>Biometrics</em>,  44:35-47.</p>
<p>Segal M.R. and Xiao Y. Multivariate random
  forests. (2011). <em>Wiley Interdisciplinary Reviews: Data Mining
  and Knowledge Discovery</em>. 1(1):80-87.</p>
<p>Tang F. and Ishwaran H. (2017).  Random forest missing data
  algorithms.  <em>Statistical Analysis and Data Mining</em>, 10:363-377.</p>
<p>Therneau, T.M. (2024).  Concordance.  Vignette in the <span class="pkg">survival</span> R
  package.</p>
<p>Uno, H., Cai, T., Pencina, M.J., D'Agostino, Ralph B. and Wei, L-J. (2011).
  On the C-statistics for evaluating overall adequacy of risk prediction procedures with censored survival data.
  <em>Statistics in Medicine</em>, 30:1105-1117.</p>
<p>Zhang H., Zimmerman J., Nettleton D. and Nordman D.J. (2019).  Random
  forest prediction intervals. <em>The American Statistician</em>. 4:1-5.</p>


    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index">

<p><code class="command"><a href="get.tree.rfsrc.html">get.tree.rfsrc</a></code>,</p>
<p><code class="command"><a href="holdout.vimp.rfsrc.html">holdout.vimp.rfsrc</a></code>,</p>
<p><code class="command"><a href="imbalanced.rfsrc.html">imbalanced.rfsrc</a></code>,
  <code class="command"><a href="impute.rfsrc.html">impute.rfsrc</a></code>,</p>
<p><code class="command"><a href="max.subtree.rfsrc.html">max.subtree.rfsrc</a></code>,</p>
<p><code class="command"><a href="partial.rfsrc.html">partial.rfsrc</a></code>,
  <code class="command"><a href="plot.competing.risk.rfsrc.html">plot.competing.risk.rfsrc</a></code>,
  <code class="command"><a href="plot.rfsrc.html">plot.rfsrc</a></code>,
  <code class="command"><a href="plot.survival.rfsrc.html">plot.survival.rfsrc</a></code>,
  <code class="command"><a href="plot.variable.rfsrc.html">plot.variable.rfsrc</a></code>,
  <code class="command"><a href="predict.rfsrc.html">predict.rfsrc</a></code>,
  <code class="command"><a href="print.rfsrc.html">print.rfsrc</a></code>,</p>
<p><code class="command"><a href="quantreg.rfsrc.html">quantreg.rfsrc</a></code>,</p>
<p><code class="command">rfsrc</code>,
  <code class="command"><a href="rfsrc.anonymous.html">rfsrc.anonymous</a></code>,
  <code class="command">rfsrc.cart</code>,
  <code class="command"><a href="rfsrc.fast.html">rfsrc.fast</a></code>,</p>

<p><code class="command"><a href="sidClustering.rfsrc.html">sidClustering.rfsrc</a></code>,</p>
<p><code class="command"><a href="subsample.rfsrc.html">subsample.rfsrc</a></code>,</p>
<p><code class="command"><a href="tune.rfsrc.html">tune.rfsrc</a></code>,</p>
<p><code class="command"><a href="vimp.rfsrc.html">vimp.rfsrc</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># \donttest{</span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span><span class="co">## survival analysis</span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## veteran data</span></span>
<span><span class="co">## randomized trial of two treatment regimens for lung cancer</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">veteran</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">v.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">veteran</span>, block.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plot tree number 3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="get.tree.rfsrc.html">get.tree</a></span><span class="op">(</span><span class="va">v.obj</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## print results of trained forest</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">v.obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plot results of trained forest</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v.obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plot survival curves for first 10 individuals -- direct way</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html" class="external-link">matplot</a></span><span class="op">(</span><span class="va">v.obj</span><span class="op">$</span><span class="va">time.interest</span>, <span class="fl">100</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">v.obj</span><span class="op">$</span><span class="va">survival.oob</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Time"</span>, ylab <span class="op">=</span> <span class="st">"Survival"</span>, type <span class="op">=</span> <span class="st">"l"</span>, lty <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plot survival curves for first 10 individuals</span></span>
<span><span class="co">## using function "plot.survival" </span></span>
<span><span class="fu"><a href="plot.survival.rfsrc.html">plot.survival</a></span><span class="op">(</span><span class="va">v.obj</span>, subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## obtain Brier score using KM and RSF censoring distribution estimators</span></span>
<span><span class="va">bs.km</span> <span class="op">&lt;-</span> <span class="fu"><a href="plot.survival.rfsrc.html">get.brier.survival</a></span><span class="op">(</span><span class="va">v.obj</span>, cens.model <span class="op">=</span> <span class="st">"km"</span><span class="op">)</span><span class="op">$</span><span class="va">brier.score</span></span>
<span><span class="va">bs.rsf</span> <span class="op">&lt;-</span> <span class="fu"><a href="plot.survival.rfsrc.html">get.brier.survival</a></span><span class="op">(</span><span class="va">v.obj</span>, cens.model <span class="op">=</span> <span class="st">"rfsrc"</span><span class="op">)</span><span class="op">$</span><span class="va">brier.score</span></span>
<span></span>
<span><span class="co">## plot the brier score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bs.km</span>, type <span class="op">=</span> <span class="st">"s"</span>, col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">bs.rsf</span>, type <span class="op">=</span><span class="st">"s"</span>, col <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cens.model = km"</span>, <span class="st">"cens.model = rfsrc"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plot CRPS (continuous rank probability score) as function of time</span></span>
<span><span class="co">## here's how to calculate the CRPS for every time point</span></span>
<span><span class="va">trapz</span> <span class="op">&lt;-</span> <span class="fu">randomForestSRC</span><span class="fu">:::</span><span class="va">trapz</span></span>
<span><span class="va">time</span> <span class="op">&lt;-</span> <span class="va">v.obj</span><span class="op">$</span><span class="va">time.interest</span></span>
<span><span class="va">crps.km</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">time</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">trapz</span><span class="op">(</span><span class="va">time</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">j</span><span class="op">]</span>, <span class="va">bs.km</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">j</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">time</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">crps.rsf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">time</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">trapz</span><span class="op">(</span><span class="va">time</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">j</span><span class="op">]</span>, <span class="va">bs.rsf</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">j</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">time</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">crps.km</span>, ylab <span class="op">=</span> <span class="st">"CRPS"</span>, type <span class="op">=</span> <span class="st">"s"</span>, col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">crps.rsf</span>, type <span class="op">=</span><span class="st">"s"</span>, col <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottomright"</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cens.model = km"</span>, <span class="st">"cens.model = rfsrc"</span><span class="op">)</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## fast nodesize optimization for veteran data</span></span>
<span><span class="co">## optimal nodesize in survival is larger than other families</span></span>
<span><span class="co">## see the function "tune" for more examples</span></span>
<span><span class="fu"><a href="tune.rfsrc.html">tune.nodesize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>,<span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">veteran</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Primary biliary cirrhosis (PBC) of the liver</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbc</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">pbc.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">days</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pbc.obj</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## save.memory example for survival</span></span>
<span><span class="co">## growing many deep trees creates memory issue without this option!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbc</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">days</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span>, splitrule <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>            ntree <span class="op">=</span> <span class="fl">25000</span>, nodesize <span class="op">=</span> <span class="fl">1</span>, save.memory <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span><span class="co">## trees can be plotted for any family </span></span>
<span><span class="co">## see get.tree for details and more examples</span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## survival where factors have many levels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">veteran</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="va">veteran</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">celltype</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">vd</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">diagtime</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">vd</span><span class="op">$</span><span class="va">diagtime</span><span class="op">)</span></span>
<span><span class="va">vd.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>,<span class="va">status</span><span class="op">)</span><span class="op">~</span><span class="va">.</span>, <span class="va">vd</span>, ntree <span class="op">=</span> <span class="fl">100</span>, nodesize <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="get.tree.rfsrc.html">get.tree</a></span><span class="op">(</span><span class="va">vd.obj</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## classification</span></span>
<span><span class="va">iris.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">Species</span> <span class="op">~</span><span class="va">.</span>, data <span class="op">=</span> <span class="va">iris</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="get.tree.rfsrc.html">get.tree</a></span><span class="op">(</span><span class="va">iris.obj</span>, <span class="fl">25</span>, class.type <span class="op">=</span> <span class="st">"bayes"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="get.tree.rfsrc.html">get.tree</a></span><span class="op">(</span><span class="va">iris.obj</span>, <span class="fl">25</span>, target <span class="op">=</span> <span class="st">"setosa"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="get.tree.rfsrc.html">get.tree</a></span><span class="op">(</span><span class="va">iris.obj</span>, <span class="fl">25</span>, target <span class="op">=</span> <span class="st">"versicolor"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="get.tree.rfsrc.html">get.tree</a></span><span class="op">(</span><span class="va">iris.obj</span>, <span class="fl">25</span>, target <span class="op">=</span> <span class="st">"virginica"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## simple example of VIMP using iris classification</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## directly from trained forest</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">rfsrc</span><span class="op">(</span><span class="va">Species</span><span class="op">~</span><span class="va">.</span>,<span class="va">iris</span>,importance<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## VIMP (and performance) use misclassification error by default</span></span>
<span><span class="co">## but brier prediction error can be requested</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">rfsrc</span><span class="op">(</span><span class="va">Species</span><span class="op">~</span><span class="va">.</span>,<span class="va">iris</span>,importance<span class="op">=</span><span class="cn">TRUE</span>,perf.type<span class="op">=</span><span class="st">"brier"</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## example using vimp function (see vimp help file for details)</span></span>
<span><span class="va">iris.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">Species</span> <span class="op">~</span><span class="va">.</span>, data <span class="op">=</span> <span class="va">iris</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="vimp.rfsrc.html">vimp</a></span><span class="op">(</span><span class="va">iris.obj</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="vimp.rfsrc.html">vimp</a></span><span class="op">(</span><span class="va">iris.obj</span>,perf.type<span class="op">=</span><span class="st">"brier"</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## example using hold out vimp (see holdout.vimp help file for details)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="holdout.vimp.rfsrc.html">holdout.vimp</a></span><span class="op">(</span><span class="va">Species</span><span class="op">~</span><span class="va">.</span>,<span class="va">iris</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="holdout.vimp.rfsrc.html">holdout.vimp</a></span><span class="op">(</span><span class="va">Species</span><span class="op">~</span><span class="va">.</span>,<span class="va">iris</span>,perf.type<span class="op">=</span><span class="st">"brier"</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## confidence interval for vimp using subsampling</span></span>
<span><span class="co">## compare with holdout vimp</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## new York air quality measurements</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">Ozone</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">airquality</span><span class="op">)</span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsample.rfsrc.html">subsample</a></span><span class="op">(</span><span class="va">o</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">so</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## compare with holdout vimp</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="holdout.vimp.rfsrc.html">holdout.vimp</a></span><span class="op">(</span><span class="va">Ozone</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">airquality</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span><span class="co">## example of imputation in survival analysis</span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbc</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">pbc.obj2</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">days</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span>, na.action <span class="op">=</span> <span class="st">"na.impute"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## same as above but iterate the missing data algorithm</span></span>
<span><span class="va">pbc.obj3</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">days</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span>,</span>
<span>         na.action <span class="op">=</span> <span class="st">"na.impute"</span>, nimpute <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## fast way to impute data (no inference is done)</span></span>
<span><span class="co">## see impute for more details</span></span>
<span><span class="va">pbc.imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="impute.rfsrc.html">impute</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">days</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span>, splitrule <span class="op">=</span> <span class="st">"random"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span><span class="co">## compare RF-SRC to Cox regression</span></span>
<span><span class="co">## Illustrates C-error and Brier score measures of performance</span></span>
<span><span class="co">## assumes "pec" and "survival" libraries are loaded</span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/therneau/survival" class="external-link">"survival"</a></span>, logical.return <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="op">&amp;</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"pec"</span>, logical.return <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="op">&amp;</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"prodlim"</span>, logical.return <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="op">{</span></span>
<span>  <span class="co">##prediction function required for pec</span></span>
<span>  <span class="va">predictSurvProb.rfsrc</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">object</span>, <span class="va">newdata</span>, <span class="va">times</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">ptemp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">object</span>,newdata<span class="op">=</span><span class="va">newdata</span>,<span class="va">...</span><span class="op">)</span><span class="op">$</span><span class="va">survival</span></span>
<span>    <span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/prodlim/man/sindex.html" class="external-link">sindex</a></span><span class="op">(</span>jump.times <span class="op">=</span> <span class="va">object</span><span class="op">$</span><span class="va">time.interest</span>, eval.times <span class="op">=</span> <span class="va">times</span><span class="op">)</span></span>
<span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">ptemp</span><span class="op">)</span><span class="op">[</span>, <span class="va">pos</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NROW</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NROW</a></span><span class="op">(</span><span class="va">newdata</span><span class="op">)</span> <span class="op">||</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NCOL</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">times</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Prediction failed"</span><span class="op">)</span></span>
<span>    <span class="va">p</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co">## data, formula specifications</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbc</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span>  <span class="va">pbc.na</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">pbc</span><span class="op">)</span>  <span class="co">##remove NA's</span></span>
<span>  <span class="va">surv.f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">days</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span>
<span>  <span class="va">pec.f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/prodlim/man/Hist.html" class="external-link">Hist</a></span><span class="op">(</span><span class="va">days</span>,<span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## run cox/rfsrc models</span></span>
<span>  <span class="co">## for illustration we use a small number of trees</span></span>
<span>  <span class="va">cox.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/coxph.html" class="external-link">coxph</a></span><span class="op">(</span><span class="va">surv.f</span>, data <span class="op">=</span> <span class="va">pbc.na</span>, x <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">rfsrc.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">surv.f</span>, <span class="va">pbc.na</span>, ntree <span class="op">=</span> <span class="fl">150</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## compute bootstrap cross-validation estimate of expected Brier score</span></span>
<span>  <span class="co">## see Mogensen, Ishwaran and Gerds (2012) Journal of Statistical Software</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">17743</span><span class="op">)</span></span>
<span>  <span class="va">prederror.pbc</span> <span class="op">&lt;-</span> <span class="fu">pec</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">cox.obj</span>,<span class="va">rfsrc.obj</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pbc.na</span>, formula <span class="op">=</span> <span class="va">pec.f</span>,</span>
<span>                        splitMethod <span class="op">=</span> <span class="st">"bootcv"</span>, B <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">prederror.pbc</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">prederror.pbc</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## compute out-of-bag C-error for cox regression and compare to rfsrc</span></span>
<span>  <span class="co">## use uno weights throughout</span></span>
<span>  <span class="va">rfsrc.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">surv.f</span>, <span class="va">pbc.na</span><span class="op">)</span></span>
<span>  <span class="va">uno.wt</span> <span class="op">&lt;-</span> <span class="va">rfsrc.obj</span><span class="op">$</span><span class="va">forest</span><span class="op">$</span><span class="va">uno.weights</span><span class="op">$</span><span class="va">weight</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"out-of-bag Cox Analysis ..."</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>  <span class="va">cox.err</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="kw">function</span><span class="op">(</span><span class="va">b</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">b</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">10</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"cox bootstrap:"</span>, <span class="va">b</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>    <span class="va">train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbc.na</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbc.na</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">cox.obj</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/coxph.html" class="external-link">coxph</a></span><span class="op">(</span><span class="va">surv.f</span>, <span class="va">pbc.na</span><span class="op">[</span><span class="va">train</span>, <span class="op">]</span><span class="op">)</span><span class="op">}</span>, error<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span><span class="op">{</span><span class="cn">NULL</span><span class="op">}</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">cox.obj</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu">get.cindex</span><span class="op">(</span><span class="va">pbc.na</span><span class="op">$</span><span class="va">days</span><span class="op">[</span><span class="op">-</span><span class="va">train</span><span class="op">]</span>,</span>
<span>                 <span class="va">pbc.na</span><span class="op">$</span><span class="va">status</span><span class="op">[</span><span class="op">-</span><span class="va">train</span><span class="op">]</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cox.obj</span>, <span class="va">pbc.na</span><span class="op">[</span><span class="op">-</span><span class="va">train</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>                 weight<span class="op">=</span><span class="va">uno.wt</span><span class="op">[</span><span class="op">-</span><span class="va">train</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="cn">NA</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n\tOOB error rates (with Uno-IPCW):\n"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\tRSF            : "</span>, <span class="fu">get.mv.error</span><span class="op">(</span><span class="va">rfsrc.obj</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\tCox regression : "</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">cox.err</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span><span class="co">## competing risks</span></span>
<span><span class="co">##------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## WIHS analysis</span></span>
<span><span class="co">## cumulative incidence function (CIF) for HAART and AIDS stratified by IDU</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">wihs</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">wihs.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">wihs</span>, nsplit <span class="op">=</span> <span class="fl">3</span>, ntree <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="plot.competing.risk.rfsrc.html">plot.competing.risk</a></span><span class="op">(</span><span class="va">wihs.obj</span><span class="op">)</span></span>
<span><span class="va">cif</span> <span class="op">&lt;-</span> <span class="va">wihs.obj</span><span class="op">$</span><span class="va">cif.oob</span></span>
<span><span class="va">Time</span> <span class="op">&lt;-</span> <span class="va">wihs.obj</span><span class="op">$</span><span class="va">time.interest</span></span>
<span><span class="va">idu</span> <span class="op">&lt;-</span> <span class="va">wihs</span><span class="op">$</span><span class="va">idu</span></span>
<span><span class="va">cif.haart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">cif</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">]</span><span class="op">[</span><span class="va">idu</span> <span class="op">==</span> <span class="fl">0</span>,<span class="op">]</span>, <span class="fl">2</span>, <span class="va">mean</span><span class="op">)</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">cif</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">]</span><span class="op">[</span><span class="va">idu</span> <span class="op">==</span> <span class="fl">1</span>,<span class="op">]</span>, <span class="fl">2</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cif.aids</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">cif</span><span class="op">[</span>,,<span class="fl">2</span><span class="op">]</span><span class="op">[</span><span class="va">idu</span> <span class="op">==</span> <span class="fl">0</span>,<span class="op">]</span>, <span class="fl">2</span>, <span class="va">mean</span><span class="op">)</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">cif</span><span class="op">[</span>,,<span class="fl">2</span><span class="op">]</span><span class="op">[</span><span class="va">idu</span> <span class="op">==</span> <span class="fl">1</span>,<span class="op">]</span>, <span class="fl">2</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html" class="external-link">matplot</a></span><span class="op">(</span><span class="va">Time</span>, <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">cif.haart</span>, <span class="va">cif.aids</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span>,</span>
<span>        lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">3</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Cumulative Incidence"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HAART (Non-IDU)"</span>, <span class="st">"HAART (IDU)"</span>, <span class="st">"AIDS (Non-IDU)"</span>, <span class="st">"AIDS (IDU)"</span><span class="op">)</span>,</span>
<span>       lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">3</span>, cex <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## illustrates the various splitting rules</span></span>
<span><span class="co">## illustrates event specific and non-event specific variable selection</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/therneau/survival" class="external-link">"survival"</a></span>, logical.return <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="co">## use the pbc data from the survival package</span></span>
<span>  <span class="co">## events are transplant (1) and death (2)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbc</span>, package <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span>  <span class="va">pbc</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span>  <span class="co">## modified Gray's weighted log-rank splitting</span></span>
<span>  <span class="co">## (equivalent to cause=c(1,1) and splitrule="logrankCR")</span></span>
<span>  <span class="va">pbc.cr</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span><span class="op">)</span></span>
<span> </span>
<span>  <span class="co">## log-rank cause-1 specific splitting and targeted VIMP for cause 1</span></span>
<span>  <span class="va">pbc.log1</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span>, </span>
<span>              splitrule <span class="op">=</span> <span class="st">"logrankCR"</span>, cause <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span>, importance <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## log-rank cause-2 specific splitting and targeted VIMP for cause 2</span></span>
<span>  <span class="va">pbc.log2</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">pbc</span>, </span>
<span>              splitrule <span class="op">=</span> <span class="st">"logrankCR"</span>, cause <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, importance <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## extract VIMP from the log-rank forests: event-specific</span></span>
<span>  <span class="co">## extract minimal depth from the Gray log-rank forest: non-event specific</span></span>
<span>  <span class="va">var.perf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>md <span class="op">=</span> <span class="fu"><a href="max.subtree.rfsrc.html">max.subtree</a></span><span class="op">(</span><span class="va">pbc.cr</span><span class="op">)</span><span class="op">$</span><span class="va">order</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>                         vimp1 <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">pbc.log1</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span> ,<span class="fl">1</span><span class="op">]</span>,</span>
<span>                         vimp2 <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">pbc.log2</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span> ,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">var.perf</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">var.perf</span><span class="op">$</span><span class="va">md</span><span class="op">)</span>, <span class="op">]</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## regression analysis</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## new York air quality measurements</span></span>
<span><span class="va">airq.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">Ozone</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">airquality</span>, na.action <span class="op">=</span> <span class="st">"na.impute"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># partial plot of variables (see plot.variable for more details)</span></span>
<span><span class="fu"><a href="plot.variable.rfsrc.html">plot.variable</a></span><span class="op">(</span><span class="va">airq.obj</span>, partial <span class="op">=</span> <span class="cn">TRUE</span>, smooth.lines <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## motor trend cars</span></span>
<span><span class="va">mtcars.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">mpg</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">mtcars</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## regression with custom bootstrap</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="va">ntree</span> <span class="op">&lt;-</span> <span class="fl">25</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mtcars</span><span class="op">)</span></span>
<span><span class="va">s.size</span> <span class="op">&lt;-</span> <span class="va">n</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="va">swr</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="va">samp</span> <span class="op">&lt;-</span> <span class="fu">randomForestSRC</span><span class="fu">:::</span><span class="fu">make.sample</span><span class="op">(</span><span class="va">ntree</span>, <span class="va">n</span>, <span class="va">s.size</span>, <span class="va">swr</span><span class="op">)</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">mpg</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">mtcars</span>, bootstrap <span class="op">=</span> <span class="st">"by.user"</span>, samp <span class="op">=</span> <span class="va">samp</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## classification analysis</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## iris data</span></span>
<span><span class="va">iris.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">Species</span> <span class="op">~</span><span class="va">.</span>, data <span class="op">=</span> <span class="va">iris</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## wisconsin prognostic breast cancer data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">breast</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">breast.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">status</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">breast</span>, block.size<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">breast.obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## big data set, reduce number of variables using simple method</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## use Iowa housing data set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">housing</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## original data contains lots of missing data, use fast imputation</span></span>
<span><span class="co">## however see impute for other methods</span></span>
<span><span class="va">housing2</span> <span class="op">&lt;-</span> <span class="fu"><a href="impute.rfsrc.html">impute</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">housing</span>, fast <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## run shallow trees to find variables that split any tree</span></span>
<span><span class="va">xvar.used</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">SalePrice</span> <span class="op">~</span><span class="va">.</span>, <span class="va">housing2</span>, ntree <span class="op">=</span> <span class="fl">250</span>, nodedepth <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                   var.used<span class="op">=</span><span class="st">"all.trees"</span>, mtry <span class="op">=</span> <span class="cn">Inf</span>, nsplit <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">$</span><span class="va">var.used</span></span>
<span></span>
<span><span class="co">## now fit forest using filtered variables</span></span>
<span><span class="va">xvar.keep</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">xvar.used</span><span class="op">)</span><span class="op">[</span><span class="va">xvar.used</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">SalePrice</span><span class="op">~</span><span class="va">.</span>, <span class="va">housing2</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SalePrice"</span>, <span class="va">xvar.keep</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">o</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## imbalanced classification data</span></span>
<span><span class="co">## see the "imbalanced" function for further details</span></span>
<span><span class="co">##</span></span>
<span><span class="co">## (a) use balanced random forests with undersampling of the majority class</span></span>
<span><span class="co">## Specifically let n0, n1 be sample sizes for majority, minority</span></span>
<span><span class="co">## cases.  We sample 2 x n1 cases with majority, minority cases chosen</span></span>
<span><span class="co">## with probabilities n1/n, n0/n where n=n0+n1</span></span>
<span><span class="co">##</span></span>
<span><span class="co">## (b) balanced random forests using "imbalanced"</span></span>
<span><span class="co">##</span></span>
<span><span class="co">## (c) q-classifier (RFQ) using "imbalanced" </span></span>
<span><span class="co">##</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## Wisconsin breast cancer example</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">breast</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span><span class="va">breast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">breast</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## balanced random forests - brute force</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">breast</span><span class="op">$</span><span class="va">status</span></span>
<span><span class="va">obdirect</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">status</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">breast</span>, nsplit <span class="op">=</span> <span class="fl">10</span>,</span>
<span>            case.wt <span class="op">=</span> <span class="fu">randomForestSRC</span><span class="fu">:::</span><span class="fu">make.wt</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,</span>
<span>            sampsize <span class="op">=</span> <span class="fu">randomForestSRC</span><span class="fu">:::</span><span class="fu">make.size</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">obdirect</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.imbalanced.performance</span><span class="op">(</span><span class="va">obdirect</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## balanced random forests - using "imbalanced" </span></span>
<span><span class="va">ob</span> <span class="op">&lt;-</span> <span class="fu"><a href="imbalanced.rfsrc.html">imbalanced</a></span><span class="op">(</span><span class="va">status</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">breast</span>, method <span class="op">=</span> <span class="st">"brf"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ob</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.imbalanced.performance</span><span class="op">(</span><span class="va">ob</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## q-classifier (RFQ) - using "imbalanced" </span></span>
<span><span class="va">oq</span> <span class="op">&lt;-</span> <span class="fu"><a href="imbalanced.rfsrc.html">imbalanced</a></span><span class="op">(</span><span class="va">status</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">breast</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">oq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.imbalanced.performance</span><span class="op">(</span><span class="va">oq</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## q-classifier (RFQ) - with auc splitting</span></span>
<span><span class="va">oqauc</span> <span class="op">&lt;-</span> <span class="fu"><a href="imbalanced.rfsrc.html">imbalanced</a></span><span class="op">(</span><span class="va">status</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">breast</span>, splitrule <span class="op">=</span> <span class="st">"auc"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">oqauc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.imbalanced.performance</span><span class="op">(</span><span class="va">oqauc</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## unsupervised analysis</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## two equivalent ways to implement unsupervised forests</span></span>
<span><span class="va">mtcars.unspv</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu">Unsupervised</span><span class="op">(</span><span class="op">)</span> <span class="op">~</span><span class="va">.</span>, data <span class="op">=</span> <span class="va">mtcars</span><span class="op">)</span></span>
<span><span class="va">mtcars2.unspv</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mtcars</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## illustration of sidClustering for the mtcars data</span></span>
<span><span class="co">## see sidClustering for more details</span></span>
<span><span class="va">mtcars.sid</span> <span class="op">&lt;-</span> <span class="fu"><a href="sidClustering.rfsrc.html">sidClustering</a></span><span class="op">(</span><span class="va">mtcars</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">mtcars</span>, <span class="va">mtcars.sid</span><span class="op">$</span><span class="va">cl</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">mtcars</span>, <span class="va">mtcars.sid</span><span class="op">$</span><span class="va">cl</span><span class="op">[</span>, <span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## bivariate regression using Mahalanobis splitting</span></span>
<span><span class="co">## also illustrates user specified covariance matrix</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"mlbench"</span>, logical.return <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="co">## load boston housing data, specify the bivariate regression</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">BostonHousing</span><span class="op">)</span></span>
<span>  <span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></span><span class="op">(</span><span class="st">"Multivar(lstat, nox) ~."</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## Mahalanobis splitting  </span></span>
<span>  <span class="va">bh.mreg</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">f</span>, <span class="va">BostonHousing</span>, importance <span class="op">=</span> <span class="cn">TRUE</span>, splitrule <span class="op">=</span> <span class="st">"mahal"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## performance error and vimp</span></span>
<span>  <span class="va">vmp</span> <span class="op">&lt;-</span> <span class="fu">get.mv.vimp</span><span class="op">(</span><span class="va">bh.mreg</span><span class="op">)</span></span>
<span>  <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">get.mv.predicted</span><span class="op">(</span><span class="va">bh.mreg</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## standardized error and vimp</span></span>
<span>  <span class="va">err.std</span> <span class="op">&lt;-</span> <span class="fu">get.mv.error</span><span class="op">(</span><span class="va">bh.mreg</span>, standardize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">vmp.std</span> <span class="op">&lt;-</span> <span class="fu">get.mv.vimp</span><span class="op">(</span><span class="va">bh.mreg</span>, standardize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## same analysis, but with user specified covariance matrix</span></span>
<span>  <span class="va">sigma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cov</a></span><span class="op">(</span><span class="va">BostonHousing</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lstat"</span>,<span class="st">"nox"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">bh.mreg2</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">f</span>, <span class="va">BostonHousing</span>, splitrule <span class="op">=</span> <span class="st">"mahal"</span>, sigma <span class="op">=</span> <span class="va">sigma</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## multivariate mixed forests (nutrigenomic study)</span></span>
<span><span class="co">## study effects of diet, lipids and gene expression for mice</span></span>
<span><span class="co">## diet, genotype and lipids used as the multivariate y</span></span>
<span><span class="co">## genes used for the x features</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## load the data (data is a list)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nutrigenomic</span>, package <span class="op">=</span> <span class="st">"randomForestSRC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## assemble the multivariate y data</span></span>
<span><span class="va">ydta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>diet <span class="op">=</span> <span class="va">nutrigenomic</span><span class="op">$</span><span class="va">diet</span>,</span>
<span>                   genotype <span class="op">=</span> <span class="va">nutrigenomic</span><span class="op">$</span><span class="va">genotype</span>,</span>
<span>                   <span class="va">nutrigenomic</span><span class="op">$</span><span class="va">lipids</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## multivariate mixed forest call</span></span>
<span><span class="co">## uses "get.mv.formula" for conveniently setting formula</span></span>
<span><span class="va">mv.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu">get.mv.formula</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ydta</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">ydta</span>, <span class="va">nutrigenomic</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span>,</span>
<span>    importance<span class="op">=</span><span class="cn">TRUE</span>, nsplit <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## print results for diet and genotype y values      </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">mv.obj</span>, outcome.target <span class="op">=</span> <span class="st">"diet"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">mv.obj</span>, outcome.target <span class="op">=</span> <span class="st">"genotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## extract standardized VIMP</span></span>
<span><span class="va">svimp</span> <span class="op">&lt;-</span> <span class="fu">get.mv.vimp</span><span class="op">(</span><span class="va">mv.obj</span>, standardize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plot standardized VIMP for diet, genotype and lipid for each gene</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">svimp</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"bisque"</span>, cex.axis <span class="op">=</span> <span class="fl">.7</span>, las <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        outline <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"standardized VIMP"</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"diet/genotype/lipid VIMP for each gene"</span><span class="op">)</span></span>
<span>  </span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## illustrates yvar.wt which sets the probability of selecting </span></span>
<span><span class="co">## the response variables in multivariate regression</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## use mtcars: add fake responses</span></span>
<span><span class="va">mult.mtcars</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">mtcars</span>, <span class="va">mtcars</span><span class="op">$</span><span class="va">mpg</span>, <span class="va">mtcars</span><span class="op">$</span><span class="va">mpg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mult.mtcars</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mtcars</span><span class="op">)</span>, <span class="st">"mpg2"</span>, <span class="st">"mpg3"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## noise up the fake responses</span></span>
<span><span class="va">mult.mtcars</span><span class="op">$</span><span class="va">mpg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">mtcars</span><span class="op">$</span><span class="va">mpg</span><span class="op">)</span></span>
<span><span class="va">mult.mtcars</span><span class="op">$</span><span class="va">mpg3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">mtcars</span><span class="op">$</span><span class="va">mpg</span><span class="op">)</span></span>
<span></span>
<span><span class="va">formula</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="fu">Multivar</span><span class="op">(</span><span class="va">mpg</span>, <span class="va">mpg2</span>, <span class="va">mpg3</span><span class="op">)</span>  <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## select 2 of the 3 responses randomly at each split with an associated weight vector.</span></span>
<span><span class="co">## choose the noisy y responses which should degrade performance</span></span>
<span><span class="va">yvar.wt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.000001</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">ytry</span> <span class="op">=</span> <span class="fl">2</span></span>
<span></span>
<span><span class="va">mult.grow</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span>formula <span class="op">=</span> <span class="va">formula</span>, data <span class="op">=</span> <span class="va">mult.mtcars</span>, ytry <span class="op">=</span> <span class="va">ytry</span>, yvar.wt <span class="op">=</span> <span class="va">yvar.wt</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">mult.grow</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.mv.error</span><span class="op">(</span><span class="va">mult.grow</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Also, compare the following two results, as they should be similar:</span></span>
<span><span class="va">yvar.wt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.0</span>, <span class="fl">00000.1</span>, <span class="fl">00000.1</span><span class="op">)</span></span>
<span><span class="va">ytry</span> <span class="op">=</span> <span class="fl">1</span></span>
<span></span>
<span><span class="va">result1</span> <span class="op">=</span> <span class="fu">rfsrc</span><span class="op">(</span>formula <span class="op">=</span> <span class="va">formula</span>, data <span class="op">=</span> <span class="va">mult.mtcars</span>, ytry <span class="op">=</span> <span class="va">ytry</span>, yvar.wt <span class="op">=</span> <span class="va">yvar.wt</span><span class="op">)</span></span>
<span><span class="va">result2</span> <span class="op">=</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">mpg</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">mtcars</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.mv.error</span><span class="op">(</span><span class="va">result1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">get.mv.error</span><span class="op">(</span><span class="va">result2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span><span class="co">## custom splitting using the pre-coded examples</span></span>
<span><span class="co">## ------------------------------------------------------------</span></span>
<span></span>
<span><span class="co">## motor trend cars</span></span>
<span><span class="va">mtcars.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">mpg</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">mtcars</span>, splitrule <span class="op">=</span> <span class="st">"custom"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## iris analysis</span></span>
<span><span class="va">iris.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="va">Species</span> <span class="op">~</span><span class="va">.</span>, data <span class="op">=</span> <span class="va">iris</span>, splitrule <span class="op">=</span> <span class="st">"custom1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## WIHS analysis</span></span>
<span><span class="va">wihs.obj</span> <span class="op">&lt;-</span> <span class="fu">rfsrc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">status</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">wihs</span>, nsplit <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  ntree <span class="op">=</span> <span class="fl">100</span>, splitrule <span class="op">=</span> <span class="st">"custom1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># }</span></span></code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.luminwin.net" class="external-link">Min Lu</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

